corehq/apps/api/resources/messaging_event/pagination.py

```diff
@@ -0,0 +1,73 @@
+import logging
+from base64 import b64encode
+from urllib.parse import urlencode
+
+from corehq.apps.api.resources.messaging_event.serializers import serialize_event
+from corehq.apps.api.resources.messaging_event.utils import get_limit_offset
+from corehq.util import reverse
+
+LAST_OBJECT_ID = "last_object_id"
+
+logger = logging.getLogger(__name__)
+
+
+def get_paged_data(query, request_params):
+    limit = get_limit_offset("limit", request_params, 20, max_value=1000)
+    objects = _get_objects(query, request_params, limit)
+    return {
+        "objects": [serialize_event(event) for event in objects],
+        "meta": {
+            "limit": limit,
+            "next": _get_cursor(objects, request_params)
+        }
+    }
+
+
+def _get_objects(query, request_params, limit):
+    """Get the list of objects to return. For pages > 1 this will
+    skip the first object in the page if it was present in the previous
+    page (as encoded in the cursor)"""
+    if request_params.is_cursor:
+        objects = list(query[:limit + 1])
+        last_id = request_params.get(LAST_OBJECT_ID)
+        try:
+            last_id = int(last_id)
+        except ValueError:
+            logger.debug("invalid last_object_id: '{last_id}")
+        else:
+            if objects[0].id == last_id:
```

Comment:
If there multiple objects having the same `date` value this may not be true and yet an object with `id == last_id` may be present in the page. Would it be worth checking all objects with `date` equal to the page start date?

Reply:
The data is sorted by `date` AND `id` so I don't think that's necessary. I'll write an additional test case for it though.

Reply:
cb02ffc2c13